### ISSUE TITLE
Cannot use Trim functions in index expressions

### ISSUE THREAD

**Issue by cpyrgas** [X-Sharp/XSharpPublic#1148]
The following code throws an Unkown error at runtime, when trying to use any of the AllTrim(), LTrim() or RTrim() functions in the index expression. Same thing in VO runs without errors and seems to give the expected results.

```
FUNCTION Start() AS VOID
LOCAL cDbf AS STRING
cDbf := "c:\dbf\testcdx"

RddSetDefault("DBFCDX")

FErase(cDbf + ".cdx")
DbCreate(cDbf, {{"FLD","C",10,0}})

DbUseArea(,,cDbf)
DbAppend()
FieldPut(1, " B")
DbAppend()
FieldPut(1, " A")
? DbCreateIndex(cDbf,"AllTrim(FLD)")
? DbCreateIndex(cDbf,"LTrim(FLD)")
? DbCreateIndex(cDbf,"RTrim(FLD)")
? DbCreateIndex(cDbf,"SubStr(FLD,2,1)") // this one works OK

DbGoTop()
? FieldGet(1)
DbGoBottom()
? FieldGet(1)

DbCloseArea()
```

**Comment by cpyrgas** [comment]
Guess the problem has to do with calculating the index key length, which becomes variable this way? It somehow seems to work in VO though, so at a minimum it needs a descriptive error message about this in the runtime exception.

**Comment by RobertvanderHulst** [comment]
Yes the key length is determined by evaluating the expression at EOF.
Using Trim() was not supported in Clipper too.
If you do that then you must make sure you pad the index to a fixed length.
Something like this:
`Left(Trim(LastName)+Trim(FirstName)+Space(50),50)`
or 
`PadR(Trim(LastName)+Trim(FirstName),50)`

I am not sure why and how this works in VO

**Comment by RobertvanderHulst** [comment]
I have changed the RDD now. It throws an error with subcode 1022  (EDB_EXPRESSION)
with the description "Index expression 'AllTrim(FLD)'results in a string with length 0".

**Comment by cpyrgas** [comment]
Robert, that's much better, but I'm afraid the developer will still do not know what the problem is. I would add some additional text in this message, something like
" - cannot use AllTrim(), LTrim(), RTrim() and similar functions resulting to variable string lengths."

**Comment by RobertvanderHulst** [comment]
Why don't you do this yourself. The text is in the Generic.resx resource file

**Comment by cpyrgas** [comment]
Will do, but didn't want to do it without asking first!

**Comment by FFF-xs** [comment]
FTR, was'nt that the reason for the "-" string operator in VO? I seem to remember having index expressions like "Fullname-Firstname", resulting in "FallerKarl_____" (underscore to simulate blanks here)

**Comment by RobertvanderHulst** [comment]
@FFF-xs the "-" operator concatenates strings and moves all the trailing spaces to the end. It does NOT make the resulting string shorter.

**Comment by FFF-xs** [comment]
I know it keeps the length, that was the reason for using it in an index expression which concatenates e.g. two string fields with varying length of content values. As for "no sense", well, that's depends. For me it's convenient for seeking - it might not be usefull for a table browse...
